lib/models/StreamSplitter.js

```diff
@@ -0,0 +1,173 @@
+const crypto = require('crypto');
+const fs = require('fs');
+const path = require('path');
+const stream = require('stream');
+
+function randomString(size) {
+  return crypto.randomBytes(size).toString('base64url').slice(0, size);
+}
+
+class InMemoryStreamSplitter extends stream.EventEmitter {
+  constructor(stream, maxSize) {
+    super();
+
+    this.buffer = [];
+    this.stream = stream;
+    this.maxSize = maxSize;
+
+    this.partial = null;
+
+    this.stream.on('error', this.handleError.bind(this));
+    this.stream.on('data', this.handleData.bind(this));
+    this.stream.on('end', this.handleEnd.bind(this));
+  }
+
+  handleError(err) {
+    this.handleEnd();
+  }
+
+  handleEnd() {
+    if (this.partial) {
+      this.partial.stream.end();
+      this.partial = null;
+    }
+  }
+
+  handleData(data) {
+    if (!this.partial) {
+      this.partial = {
+        stream: new stream.PassThrough(),
+        bytes: 0,
+      };
+      this.emit('stream', this.partial.stream);
+    }
+
+    const overflow = this.partial.bytes + data.length - this.maxSize;
+    if (overflow > 0) {
+      this.handleData(data.splice(0, data.length - overflow));
+      this.handleData(data.splice(data.length - overflow, data.length));
+      return;
+    }
+
+    this.partial.bytes += data.length;
+    this.partial.stream.write(data);
+
+    if (overflow === 0) {
+      this.partial.stream.end();
+      this.partial = null;
+    }
+  }
+}
+
+class FileStreamSplitter extends stream.Writable {
+  constructor(options, { maxChunkSize, directory }) {
+    super(options);
+
+    this.maxChunkSize = maxChunkSize;
+
+    this.currentChunkPath = null;
+    this.currentChunkSize = null;
+    this.fd = null;
+
+    this.directory = directory;
+    this.filenameTemplate = randomString(10);
+
+    this.part = 0;
+
+    this.on('error', this._finishChunk.bind(this));
+
+  }
+
+  _write(chunk, encoding, callback) {
+    Promise.resolve()
```

Comment:
I'm inclined to start using `async`/`await` everywhere and moving away from `bind`. I'll likely get to this in the TS rewrite, but it would already help if this big rewrite sets the pace for that. What do you think?

Reply:
I reluctant to change anything at this points as I'm not able to test this thoroughly. I'm using GCS as my datastore. 
Otherwise I agree with async/await.

Reply:
Good point. I'll make setting up a S3 test suite my next to do for tus-node-server. My time is limited on this project though, so will take a bit of time.